modules/imgcodecs/src/grfmt_tiff.cpp

```diff
@@ -1424,6 +1431,169 @@ bool  TiffEncoder::write( const Mat& img, const std::vector<int>& params)
+                {
+                    pDst [ x * 3 + 0 ] = pSrc [ x * 3 + 2 ] ;
+                    pDst [ x * 3 + 1 ] = pSrc [ x * 3 + 1 ] ;
+                    pDst [ x * 3 + 2 ] = pSrc [ x * 3 + 0 ] ;
+                }
+                break;
+            case COLOR_BGRA2RGBA: //  COLOR_RGBA2BGRA:
+                for (int x = 0; x < width; ++ x)
+                {
+                    pDst [ x * 4 + 0 ] = pSrc [ x * 4 + 2 ] ;
+                    pDst [ x * 4 + 1 ] = pSrc [ x * 4 + 1 ] ;
+                    pDst [ x * 4 + 2 ] = pSrc [ x * 4 + 0 ] ;
+                    pDst [ x * 4 + 3 ] = pSrc [ x * 4 + 3 ] ;
+                }
+                break;
+
+            default:
+                CV_Assert(0);
+            break;
+            }
+        }
+        break;
+
+        case CV_32S:
+        {
+            int32_t *pSrc = (int32_t*) src.ptr(y);
+            int32_t *pDst = (int32_t*) dst.ptr(y);
+
+            switch (code)
+            {
+            case COLOR_BGR2RGB: //  COLOR_RGB2BGR:
+                for (int x = 0; x < width; ++ x)
+                {
+                    pDst [ x * 3 + 0 ] = pSrc [ x * 3 + 2 ] ;
+                    pDst [ x * 3 + 1 ] = pSrc [ x * 3 + 1 ] ;
+                    pDst [ x * 3 + 2 ] = pSrc [ x * 3 + 0 ] ;
+                }
+                break;
+            case COLOR_BGRA2RGBA: //  COLOR_RGBA2BGRA:
+                for (int x = 0; x < width; ++ x)
+                {
+                    pDst [ x * 4 + 0 ] = pSrc [ x * 4 + 2 ] ;
+                    pDst [ x * 4 + 1 ] = pSrc [ x * 4 + 1 ] ;
+                    pDst [ x * 4 + 2 ] = pSrc [ x * 4 + 0 ] ;
+                    pDst [ x * 4 + 3 ] = pSrc [ x * 4 + 3 ] ;
+                }
+                break;
+
+            default:
+                CV_Assert(0);
+            break;
+            }
+        }
+        break;
+
+        case CV_64F:
+        {
+            double *pSrc = (double*) src.ptr(y);
+            double *pDst = (double*) dst.ptr(y);
+
+            switch (code)
+            {
+            case COLOR_BGR2RGB: //  COLOR_RGB2BGR:
+                for (int x = 0; x < width; ++ x)
+                {
+                    pDst [ x * 3 + 0 ] = pSrc [ x * 3 + 2 ] ;
+                    pDst [ x * 3 + 1 ] = pSrc [ x * 3 + 1 ] ;
+                    pDst [ x * 3 + 2 ] = pSrc [ x * 3 + 0 ] ;
+                }
+                break;
+            case COLOR_BGRA2RGBA: //  COLOR_RGBA2BGRA:
+                for (int x = 0; x < width; ++ x)
+                {
+                    pDst [ x * 4 + 0 ] = pSrc [ x * 4 + 2 ] ;
+                    pDst [ x * 4 + 1 ] = pSrc [ x * 4 + 1 ] ;
+                    pDst [ x * 4 + 2 ] = pSrc [ x * 4 + 0 ] ;
+                    pDst [ x * 4 + 3 ] = pSrc [ x * 4 + 3 ] ;
+                }
+                break;
+
+            default:
+                CV_Assert(0);
+            break;
+            }
+        }
+        break;
+
+        default:
+            CV_Assert(0);
```

Comment:
Please add meaningful message to the assertion.

Reply:
Thank you for your comment! 
` CV_Assert(0 && "Meaningful message" ) ` seems better. 

Current PR removes copying from SRC to DST. So this problem will not contains.